hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java

```diff
@@ -0,0 +1,272 @@
+                bufferObjectDataInputFunc,
+                bufferObjectDataOutputSupplier);
+    }
+
+    public GenericRecordBuilder createGenericRecordCloner(Schema schema, CompactInternalGenericRecord record) {
+        return new SerializingGenericRecordCloner(this, schema, record,
+                bufferObjectDataInputFunc,
+                bufferObjectDataOutputSupplier);
+    }
+
+    //========================== WRITE =============================//
+    @Override
+    public void write(ObjectDataOutput out, Object o) throws IOException {
+        assert out instanceof BufferObjectDataOutput;
+        BufferObjectDataOutput bufferObjectDataOutput = (BufferObjectDataOutput) out;
+        write(bufferObjectDataOutput, o, false);
+    }
+
+    void write(BufferObjectDataOutput out, Object o, boolean includeSchemaOnBinary) throws IOException {
+        if (o instanceof CompactGenericRecord) {
+            writeGenericRecord(out, (CompactGenericRecord) o, includeSchemaOnBinary);
+        } else {
+            writeObject(out, o, includeSchemaOnBinary);
+        }
+    }
+
+    void writeGenericRecord(BufferObjectDataOutput output, CompactGenericRecord record,
+                            boolean includeSchemaOnBinary) throws IOException {
+        Schema schema = record.getSchema();
+        if (!schema.isSchemaIdSet()) {
+            long schemaId = calculateSchemaId(schema);
+            schema.setSchemaId(schemaId);
+        }
+        schemaService.put(schema);
+        writeSchema(output, includeSchemaOnBinary, schema);
+        DefaultCompactWriter writer = new DefaultCompactWriter(this, output, schema, includeSchemaOnBinary);
+        Collection<FieldDescriptor> fields = schema.getFields();
+        for (FieldDescriptor fieldDescriptor : fields) {
+            String fieldName = fieldDescriptor.getFieldName();
+            FieldType fieldType = fieldDescriptor.getType();
+            fieldOperations(fieldType).readFromGenericRecordToWriter(writer, record, fieldName);
+        }
+        writer.end();
+    }
+
+    private long calculateSchemaId(Schema schema) throws IOException {
+        BufferObjectDataOutput out = bufferObjectDataOutputSupplier.get();
+        schema.writeData(out);
+        return RabinFingerPrint.fingerprint64(out.toByteArray());
+    }
+
+    public void writeObject(BufferObjectDataOutput out, Object o, boolean includeSchemaOnBinary) throws IOException {
+        ConfigurationRegistry registry = getOrCreateRegistry(o);
+        Class<?> aClass = o.getClass();
+
+        Schema schema = classToSchemaMap.get(aClass);
+        if (schema == null) {
+            SchemaWriter writer = new SchemaWriter(registry.getTypeName());
+            registry.getSerializer().write(writer, o);
+            schema = writer.build();
+            long schemaId = calculateSchemaId(schema);
+            schema.setSchemaId(schemaId);
+            schemaService.put(schema);
+            classToSchemaMap.put(aClass, schema);
```

Comment:
Whatever they have corresponding to the class on the java. This is a local map meaning it will not be looked up by a remote client/server. 
I don't think that it needs to be common for all languages.
Any identifier which uniquely identifies a class/struct works. 

In which path you think that it should be common for all languages?

Reply:
you use it while calculating the schema id from binary, and the binary includes the class name, hence it needs to be calculated to the exact same schema id for all languages.

Reply:
Name is already user provided. 
We have `CompactSerializerConfig.register(Class<T> clazz, String typeName);`
If it is not given we use clazz.getClassName();
But it is irrelevant what we use as a key to this map. That is why I got confused.

Reply:
If we allow implicit class name usage for Java and if the user later on in his/her application wants to add a non-java client, it won't work, this may not be a good user experience. If this name even if it is implicitly deducted, if it could be generated the same name from every language, that would for example work, but what would be the name? An idea, may be we can just use the schema id converted to string, and give the user ability to obtain a schema id for a type. Hence, we do not use java class name at all.

Reply:
It won't work in the following situation for java and zero-config case.

I have two clients with exact same set of classes in the local. 
- First client puts the schema and data to the cluster. 
- Second client gets the data before doing any put. Also retrieves the schema from the cluster because there are no schemas on the local yet. 
- On the schema, we haven't written any class name. So the second client can not create the class which actually exists in its local. After first put, it will start to return the object. I think it is important to make the return type consistent regardless of in which state the client is.

Reply:
And I don't see the following more convenient than typing Java class name. 
`give the user ability to obtain a schema id for a type.`

For the schema file first case, we will not have the Java class name written anyway. I think the current design is better.

Reply:
schema id string may not be a good solution. i am just not happy forcing a non-java user for example a python user to know and use java class naming, all of a sudden that language user will feel like java is first class citizen and i definitely don't like this. But I do not have a good solution that would fit both sides yet.